redis/src/main/java/org/hibernate/ogm/datastore/redis/AbstractRedisDialect.java

```diff
@@ -316,4 +358,126 @@ protected void storeAssociation(
 			}
 		}
 	}
+
+	/**
+	 * Scan over keys. This method is aware whether the the client is connected to a Redis Cluster.
+	 * If so, then a Redis Cluster scan requires to iterate over master nodes and keep the
+	 * state within a custom {@link KeyScanCursor} instance.
+	 *
+	 * @param cursor
+	 * @param scanArgs
+	 *
+	 * @return
+	 */
+	protected KeyScanCursor<String> scan(KeyScanCursor<String> cursor, ScanArgs scanArgs) {
+
+		if ( !clusterMode ) {
+			return scan( connection, cursor, scanArgs );
+		}
+
+		return clusterScan( cursor, scanArgs );
+	}
+
+	private KeyScanCursor<String> scan(
+			RedisClusterCommands<String, String> commands,
+			KeyScanCursor<String> cursor,
+			ScanArgs scanArgs) {
+		if ( cursor != null ) {
+			return commands.scan( cursor, scanArgs );
+		}
+		return commands.scan( scanArgs );
+	}
+
+	@SuppressWarnings("unchecked")
+	private KeyScanCursor<String> clusterScan(KeyScanCursor<String> cursor, ScanArgs scanArgs) {
+
+		RedisAdvancedClusterCommands<String, String> commands = (RedisAdvancedClusterCommands<String, String>) connection;
+		List<String> nodeIds;
+		String currentNodeId;
+
+		if ( cursor == null ) {
+			Set<RedisClusterNode> masterNodes = commands.masters().asMap().keySet();
+			nodeIds = new ArrayList<>();
+
+			for ( RedisClusterNode masterNode : masterNodes ) {
+				nodeIds.add( masterNode.getNodeId() );
+			}
+
+			if ( nodeIds.isEmpty() ) {
+				return scan( connection, cursor, scanArgs );
+			}
+
+			currentNodeId = nodeIds.get( 0 );
+		}
+		else {
+			if ( !( cursor instanceof ClusterwideKeyScanCursor ) ) {
+				throw new IllegalArgumentException(
```

Comment:
That's only an additional guard. It _should_ not happen but you know how it goes with _should_.
I'll change the Exception handling. The dialect API is not intended for outside use the guard is not necessary.